### PR TITLE
Redefine the trig functions to divide circles into 1.0 turns

### LINKED ISSUE #1059: Trigonometry functions depend oddly on the fixed-point Q setting
A circle can be divided into 360 degrees, 2*π* or *τ* radians, 1 [turn](https://en.wikipedia.org/wiki/Turn_(angle)), 2^*n* [binary radians](https://en.wikipedia.org/wiki/Binary_angular_measurement), or other units.

Currently rgbasm uses 2^*Q* binary radians, where *Q* is the fixed-point decimal precision set by `opt Q`. However, note that 2^*Q* is meant as a fixed-point value itself. For example, with the default Q16.16, a circle is divided into 2^16 = 65536.0 = `$1_0000_0000` units, which is not the same as `2**16 == 65536 == $1_0000`. With `opt Q.8`, aka Q24.8 fixed-point values, circles have 2^8 = 256.0 = `$1_00_00` units, which is not the same as `2**8 == 256 == $1_00`. 

This should be documented appropriately, since right now rgbasm(5) just claims that the trig functions "are defined in terms of a circle divided into 65535.0 degrees". It's arguably a design flaw, since the interpretation of expressions like `sin(256.0)` depends on the current `Q` setting, but changing that would break backwards compatibility.

(Ideally I'd like trig functions in turns, so `0.5` = *π* radians and `1.0` = *τ* radians no matter what. Then `sin(0.25) == 1.0` would always be true, even though the particular scale of those values would depend on `opt Q`.) 

Input:

```
MACRO sine
  println strfmt("sin(%f) = %f", (\1), sin(\1))
ENDM

 println "Q16.16:"
 opt Q.16
 def x = $0000_0000
 rept 9 ; can't use `for` since $9000_0000 is negative
   sine x
   def x += $1000_0000
 endr
 println

 println "Q24.8:"
 opt Q.8
 for x, $00_00, $90_00, $10_00
   sine x
 endr
 println

 println "Q28.4:"
 opt Q.4
 for x, $0_0, $9_0, $1_0
   sine x
 endr
 println
```

Output:

```
Q16.16:
sin(0.00000) = 0.00000
sin(4096.00000) = 0.38269
sin(8192.00000) = 0.70711
sin(12288.00000) = 0.92387
sin(16384.00000) = 1.00000
sin(20480.00000) = 0.92387
sin(24576.00000) = 0.70711
sin(28672.00000) = 0.38269
sin(32768.00000) = 0.00000

Q24.8:
sin(0.00000) = 0.00000
sin(16.00000) = 0.38281
sin(32.00000) = 0.70703
sin(48.00000) = 0.92578
sin(64.00000) = 1.00000
sin(80.00000) = 0.92578
sin(96.00000) = 0.70703
sin(112.00000) = 0.38281
sin(128.00000) = 0.00000

Q28.4:
sin(0.00000) = 0.00000
sin(1.00000) = 0.37500
sin(2.00000) = 0.68750
sin(3.00000) = 0.93750
sin(4.00000) = 1.00000
sin(5.00000) = 0.93750
sin(6.00000) = 0.68750
sin(7.00000) = 0.37500
sin(8.00000) = 0.00000
```

### PR DESCRIPTION
**Pull request by Rangi42**
This makes their behavior consistent across `Q` settings, although it breaks compatibility with the previous behavior.

Anyone using the old trig functions with Q16.16 fixed-point can get the old behavior by applying a factor of 65536 themselves, since there were 65536 "fixed-point degrees" per turn.

Fixes #1059

### CODE REVIEW

**Comment by Rangi42** [comment]
The tests are failing because the pret repos no longer match with this new `sin()` behavior. For the same reason, they currently print various new warnings. When 0.6.0 is released, pret will be updated to use it, getting rid of the warnings and errors.

File: test/asm/trigonometry.asm [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	endc
+endr
+
+SECTION "sine", ROM0[0]
+OPT Q.16
+; Generate a table of sine values from sin(0.0) to sin(1.0), with
+; amplitude scaled from [-1.0, 1.0] to [0.0, 128.0]
+DEF turns = 0.0
+REPT 256
+    db (MUL(64.0, SIN(turns)) + 64.0) >> 16
```
Comment (Rangi42) [inline test/asm/trigonometry.asm:28]: Generates the exact same numbers, but without repeating the factor of 64.
```suggestion
    db (MUL(64.0, SIN(turns) + 1.0)) >> 16
```

File: man/rgbasm.5 [hunk 2]
```diff
@@ -350,18 +350,18 @@ The trigonometry functions (
...
-ANGLE = 0.0
-    REPT 256
-        db (MUL(64.0, SIN(ANGLE)) + 64.0) >> 16
-ANGLE = ANGLE + 256.0 ; 256.0 = 65536 degrees / 256 entries
-    ENDR
+; Generate a table of sine values from sin(0.0) to sin(1.0), with
+; amplitude scaled from [-1.0, 1.0] to [0.0, 128.0]
+DEF turns = 0.0
+REPT 256
+    db (MUL(64.0, SIN(turns)) + 64.0) >> 16
```
Comment (Rangi42) [inline man/rgbasm.5:362]: Generates the exact same numbers, but without repeating the factor of 64.
```suggestion
    db (MUL(64.0, SIN(turns) + 1.0)) >> 16
```

File: src/asm/fixpoint.c [hunk 3]
```diff
@@ -24,9 +24,9 @@
 #define fix2double(i)	((double)((i) / fix_PrecisionFactor()))
 #define double2fix(d)	((int32_t)round((d) * fix_PrecisionFactor()))
 
-// pi*2 radians == 2**fixPrecision fixed-point "degrees"
-#define fdeg2rad(f)	((f) * (M_PI * 2) / fix_PrecisionFactor())
-#define rad2fdeg(r)	((r) * fix_PrecisionFactor() / (M_PI * 2))
+// pi*2 radians == 1.0 fixed-point turns
```
Comment (Rangi42) [inline src/asm/fixpoint.c:27]:
```suggestion
// 2*pi radians == 1 turn
```

File: man/rgbasm.5 [hunk 4]
```diff
@@ -350,18 +350,18 @@ The trigonometry functions (
...
-ANGLE = 0.0
-    REPT 256
-        db (MUL(64.0, SIN(ANGLE)) + 64.0) >> 16
-ANGLE = ANGLE + 256.0 ; 256.0 = 65536 degrees / 256 entries
-    ENDR
+; Generate a table of sine values from sin(0.0) to sin(1.0), with
+; amplitude scaled from [-1.0, 1.0] to [0.0, 128.0]
+DEF turns = 0.0
+REPT 256
+    db (MUL(64.0, SIN(turns) + 1.0)) >> 16
```
Comment (Rangi42) [inline man/rgbasm.5:362]:
```suggestion
    db MUL(64.0, SIN(turns) + 1.0) >> 16
```

File: test/asm/trigonometry.asm [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	endc
+endr
+
+SECTION "sine", ROM0[0]
+OPT Q.16
+; Generate a table of sine values from sin(0.0) to sin(1.0), with
+; amplitude scaled from [-1.0, 1.0] to [0.0, 128.0]
+DEF turns = 0.0
+REPT 256
+    db (MUL(64.0, SIN(turns) + 1.0)) >> 16
```
Comment (Rangi42) [inline test/asm/trigonometry.asm:28]:
```suggestion
    db MUL(64.0, SIN(turns) + 1.0) >> 16
```

**Comment by Rangi42** [comment]
If you are using trig functions in rgbasm 0.5.x, just shifting the values by 16 will correct them for 0.6.0:

| Expression | rgbasm 0.5.2 | rgbasm 0.6.0 | Notes |
| --- | --- | --- | --- |
| `sin(0.0)` | `$0` = `0.0` | `$0` = `0.0` | sin(0) = 0 |
| `sin(0.25)` | `$2` = `0.00003` | `$10000` = `1.0` | `0.25` = `1.0 / 4`, and sin(*τ* / 4) = sin(*π* / 2) = 1 |
| `sin(16384.0)` | `$10000` = `1.0` | `$0` = `0.0` | `16384.0` = `0.25 << 16` |
| `asin(1.0)` | `$4000_0000` = `16384.0` | `$4000` = `0.25` | asin(1) = *τ* / 4 = *π* / 2, and `0.25` = `16384.0 >> 16` |
| `sin($1)` | `$0` = `0.0` | `$6` = `0.000092` | `$1` = `0.00002` |
| `sin(1.0)` | `$6` = `0.000092` | `$0` = `0.0` | `1.0` = `$1 << 16` |

**Comment by ISSOtm** [comment]
So, just to be clear, it's the *input* values that must be shifted, not the *results*?

**Review by ISSOtm — APPROVED** [review]

**Comment by Rangi42** [comment]
> So, just to be clear, it's the _input_ values that must be shifted, not the _results_?

Correct. The results are unitless and thus unchanged, e.g. the sine of a quarter-turn (90 degree angle, or pi/2 radians) is 1. Only the angle unit has changed.

**Pull request merged**